[material-ui] Add new Next.js integration package (#39947)
mui/material-ui: docs/data/material/guides/nextjs/nextjs.md

### Intalling the dependencies

You should already have `@mui/material` and `next` installed. Then, install these dependencies:

<codeblock storageKey="package-manager">

```bash npm
npm install @mui/material-nextjs @emotion/cache
```

```bash yarn
yarn add @mui/material-nextjs @emotion/cache
```

```bash pnpm
pnpm add @mui/material-nextjs @emotion/cache
```

</codeblock>

### Configuring

Inside `app/layout.tsx`, import the `AppRouterCacheProvider` and wrap all of the elements under the `<body>`:

```diff
 // app/layout.tsx
+ import { AppRouterCacheProvider } from '@mui/material-nextjs/v13-appRouter';

 export default function RootLayout(props) {
   const { children } = props;
   return (
     <html lang="en">
       <body>
+        <AppRouterCacheProvider>{children}</AppRouterCacheProvider>
       </body>
     </html>
   );
 }
```

### Using other styling solutions

If you are using a styling solution other than Emotion to customize Material UI components, set `enableCssLayer` to the `options` prop:

```js
<AppRouterCacheProvider options={{ enableCssLayer: true }}>
```

> The styles generated by Material UI will be wrapped in a CSS `@layer mui` rule, which gives them precedence over custom styles—for example if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS.

Author: Oops, my bad. It should be the opposite.

# Next.js Integration

<p class="description">Learn how to use Material UI with the Next.js</p>

## App Router

[App router](https://nextjs.org/docs/app) is an evolution of the [Pages Router](#pages-router) and a new default way of building Next.js applications starting from version 13.

> This section will help you integrate Material UI with the App Router in Next.js.

Reviewer: ```suggestion
This section walks through the Material UI integration with the Next.js App Router.
```

### Intalling the dependencies

> You should already have `@mui/material` and `next` installed. Then, install these dependencies:

Reviewer: ```suggestion
Start by ensuring that you already have `@mui/material` and `next` installed. 
Then, install these dependencies:
```

<codeblock storageKey="package-manager">

```bash npm
npm install @mui/material-nextjs @emotion/cache
```

```bash yarn
yarn add @mui/material-nextjs @emotion/cache
```

```bash pnpm
pnpm add @mui/material-nextjs @emotion/cache
```

</codeblock>

### Configuring

Inside `app/layout.tsx`, import the `AppRouterCacheProvider` and wrap all of the elements under the `<body>`:

```diff
 // app/layout.tsx
+ import { AppRouterCacheProvider } from '@mui/material-nextjs/v13-appRouter'; // or `v14-appRouter` if you are using Next.js v14

 export default function RootLayout(props) {
   const { children } = props;
   return (
     <html lang="en">
       <body>
+        <AppRouterCacheProvider>{children}</AppRouterCacheProvider>
       </body>
     </html>
   );
 }
```

### Using other styling solutions

> If you are using a styling solution other than Emotion to customize Material UI components, set `enableCssLayer` to the `options` prop:

Author: Yes, thanks!

```js
<AppRouterCacheProvider options={{ enableCssLayer: true }}>
```

> This option ensures that the styles generated by Material UI will be wrapped in a CSS `@layer mui` rule, which will be overridden by anonymous layer styles—for example, if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS without using `@layer`.

Reviewer 2: is there any other way without `@layer`? we might also need to update [this list](https://mui.com/material-ui/getting-started/supported-platforms/#browser)

Author: I don't see any other way because Next.js does not provide us a way to insert the style at a specific location like in pages router.

> This option ensures that the styles generated by Material UI will be wrapped in a CSS `@layer mui` rule, which will be overridden by anonymous layer styles—for example, if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS without using `@layer`.

Reviewer: ```suggestion
This option ensures that the styles generated by Material UI will be wrapped in a CSS `@layer mui` rule, which is overridden by anonymous layer styles, in the case of using Material UI with CSS modules, Tailwind CSS, or even plain CSS without using `@layer`.
```

> To learn more about the CSS layer, see the [MDN CSS layer documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer).

Reviewer: ```suggestion
To learn more about the CSS layer, see [the MDN CSS layer documentation](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer).
```

## Pages Router

> The steps below will help you integrate Material UI with the [Pages Router](https://nextjs.org/docs/pages/building-your-application) in Next.js for both [SSR](https://nextjs.org/docs/pages/building-your-application/rendering/server-side-rendering) (server-side rendering) and [SSG](https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation) (static site generation).

Reviewer: ```suggestion
The section below walks through the Material UI integration with the Next.js [Pages Router](https://nextjs.org/docs/pages/building-your-application), for both [Server Side Rendering](https://nextjs.org/docs/pages/building-your-application/rendering/server-side-rendering) (server-side rendering) and [Static Site Generation](https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation).
```

I think it's helpful to say what the acronym stands for in the very first time it appears on the page, regardless of more experienced devs already knowing it!

### Intalling the dependencies

> You should already have `@mui/material` and `next` installed. Then, install these dependencies:

Reviewer: ```suggestion
Start by ensuring that you already have `@mui/material` and `next` installed. 
Then, install these dependencies:
```

```

```bash pnpm
pnpm add @mui/material-nextjs @emotion/cache @emotion/server
```

</codeblock>

### Configuring

Inside `pages/_document.tsx`:

- import the `documentGetInitialProps` and use it as the Document's getInitialProps.
- import the `DocumentHeadTags` and render it inside the `<Head>`.

```diff
+ import { DocumentHeadTags, documentGetInitialProps } from '@mui/material-nextjs/v13-pagesRouter'; // or `v14-pagesRouter` if you are using Next.js v14

 export default function MyDocument(props) {
   return (
     <Html lang="en">
       <Head>
+        <DocumentHeadTags {...props} />
         ...
       </Head>
       <body>
         <Main />
         <NextScript />
       </body>
     </Html>
   );
 }

+ MyDocument.getInitialProps = documentGetInitialProps;
```

Then inside `pages/_app.tsx`, import the `AppCacheProvider` and render it as the root element:

```diff
+ import { AppCacheProvider } from '@mui/material-nextjs/v13-pagesRouter'; // or `v14-pages` if you are using Next.js v14

 export default function MyApp(props) {
   return (
+    <AppCacheProvider {...props}>
       <Head>
         ...
       </Head>
       ...
+    </AppCacheProvider>
   );
 }
```

### TypeScript

> If you are using TypeScript, you will need to add `DocumentHeadTagsProps` to the Document's props interface:

Reviewer: ```suggestion
If you are using TypeScript, add `DocumentHeadTagsProps` to the Document's props interface:
```

> # Next.js Integration

Reviewer 3: Very clean guide 👌

This section will help you integrate Material UI with the App Router in Next.js.

### Intalling the dependencies

You should already have `@mui/material` and `next` installed. Then, install these dependencies:

<codeblock storageKey="package-manager">

```bash npm
npm install @mui/material-nextjs @emotion/cache
```

```bash yarn
yarn add @mui/material-nextjs @emotion/cache
```

```bash pnpm
pnpm add @mui/material-nextjs @emotion/cache
```

</codeblock>

### Configuring

Inside `app/layout.tsx`, import the `AppRouterCacheProvider` and wrap all of the elements under the `<body>`:

```diff
 // app/layout.tsx
+ import { AppRouterCacheProvider } from '@mui/material-nextjs/v13-appRouter';

 export default function RootLayout(props) {
   const { children } = props;
   return (
     <html lang="en">
       <body>
+        <AppRouterCacheProvider>{children}</AppRouterCacheProvider>
       </body>
     </html>
   );
 }
```

### Using other styling solutions

If you are using a styling solution other than Emotion to customize Material UI components, set `enableCssLayer` to the `options` prop:

```js
<AppRouterCacheProvider options={{ enableCssLayer: true }}>
```

> The styles generated by Material UI will be wrapped in a CSS `@layer mui` rule, which will be overriden by anonymous layer styles—for example if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS without using `@layer`.

Reviewer 3: ```suggestion
This option ensures that the styles generated by Material UI will be wrapped in a CSS `@layer mui` rule, which will be overridden by anonymous layer styles—for example, if you are customizing Material UI with CSS modules, Tailwind CSS, or even plain CSS without using `@layer`.
```

<codeblock storageKey="package-manager">

```bash npm
npm install @mui/material-nextjs @emotion/cache @emotion/server
```

```bash yarn
yarn add @mui/material-nextjs @emotion/cache @emotion/server
```

```bash pnpm
pnpm add @mui/material-nextjs @emotion/cache @emotion/server
```

</codeblock>

### Configuring

Inside `pages/_document.tsx`:

- import the `documentGetInitialProps` and use it as the Document's getInitialProps.
- import the `DocumentHeadTags` and render it inside the `<Head>`.

```diff
+ import { DocumentHeadTags, documentGetInitialProps } from '@mui/material-nextjs/v13-pagesRouter';

 export default function MyDocument(props) {
   return (
     <Html lang="en">
       <Head>
+        <DocumentHeadTags {...props} />
         ...
       </Head>
       <body>
         <Main />
         <NextScript />
       </body>
     </Html>
   );
 }

+ MyDocument.getInitialProps = documentGetInitialProps;
```

Then inside `pages/_app.tsx`, import the `AppCacheProvider` and render it as the root element:

```diff
+ import { AppCacheProvider } from '@mui/material-nextjs/pagesRouterV13';

 export default function MyApp(props) {

>    const { Component } = props;

Reviewer 3: ```suggestion
```

The `Component` destructed from props is not used anywhere, we can omit it for simplicity.
